rust/lance-core/src/utils/deletion.rs

```diff
@@ -115,6 +123,64 @@ impl DeletionVector {
     }
 }
 
+/// Maps a naive offset into a fragment to the local row offset that is
+/// not deleted.
+///
+/// For example, if the deletion vector is [0, 1, 2], then the mapping
+/// would be:
+///
+/// - 0 -> 3
+/// - 1 -> 4
+/// - 2 -> 5
+///
+/// and so on.
+///
+/// This expects a monotonically increasing sequence of input offsets. State
+/// is re-used between calls to `map_offset` to make the mapping more efficient.
+pub struct OffsetMapper {
+    dv: Arc<DeletionVector>,
+    left: u32,
+    last_diff: u32,
+}
+
+impl OffsetMapper {
+    pub fn new(dv: Arc<DeletionVector>) -> Self {
+        Self {
+            dv,
+            left: 0,
+            last_diff: 0,
+        }
+    }
+
+    pub fn map_offset(&mut self, offset: u32) -> u32 {
```

Comment:
I see we also have a judgment in `fragement.rs` [here](https://github.com/lancedb/lance/blob/main/rust/lance/src/dataset/fragment.rs#L2053) (but missed the judgment of discrete offset). Can we find one way to converge these judgments into one place?

Reply:
I could probably re-implement that other function in terms of this one, if we wanted. Though not sure if it is worth while.